File: solutions/python/change/1/change.py
def find_fewest_coins(coins, target):
    if target < 0:
        raise ValueError("target can't be negative")
    if target == 0:
        return []
    INF = float('inf')
    dp = [INF] * (target + 1)
    dp[0] = 0
    last = [None] * (target + 1)
    for coin in coins:
        for amount in range(coin, target + 1):
            if dp[amount - coin] + 1 < dp[amount]:
                dp[amount] = dp[amount - coin] + 1
                last[amount] = coin
    if dp[target] == INF:
        raise ValueError("can't make target with given coins")
    result = []
    while target > 0:
        result.append(last[target])
        target -= last[target]
    return sorted(result)

File: solutions/javascript/change/1/change.js
export class Change {
  calculate(coinArray, target) {
    if (target < 0) throw new Error('Negative totals are not allowed.');
    if (target === 0) return [];
    const sorted = [...coinArray].filter(c => c > 0).sort((a, b) => a - b);
    const dp = new Array(target + 1).fill(null);
    dp[0] = [];
    for (let i = 1; i <= target; i++) {
      let best = null;
      for (const coin of sorted) {
        if (coin > i) break;
        const prev = dp[i - coin];
        if (prev !== null) {
          const candidate = [...prev, coin];
          if (best === null || candidate.length < best.length) {
            best = candidate;
          } else if (candidate.length === best.length) {
            const sumDiff = coin - best[best.length - 1];
            if (sumDiff > 0) {
              let useCandidate = false;
              for (let k = best.length - 1; k >= 0; k--) {
                if (candidate[k] !== best[k]) {
                  if (candidate[k] < best[k]) useCandidate = true;
                  break;
                }
              }
              if (useCandidate) best = candidate;
            }
          }
        }
      }
      dp[i] = best;
    }
    if (dp[target] === null) {
      throw new Error(`The total ${target} cannot be represented in the given currency.`);
    }
    return dp[target].sort((a, b) => a - b);
  }
}
